pkg/block/indexheader/parallel_bucket.go

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package indexheader
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+	"github.com/thanos-io/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+	"golang.org/x/sync/errgroup"
+)
+
+// partitionSize is used for splitting range reads.
+const partitionSize = 16 * 1024 * 1024 // 16 MiB
+
+type parallelBucketReader struct {
+	objstore.BucketReader
+	tmpDir        string
+	partitionSize int64
+}
+
+func WrapWithParallel(b objstore.BucketReader, tmpDir string) objstore.BucketReader {
+	return &parallelBucketReader{
+		BucketReader:  b,
+		tmpDir:        tmpDir,
+		partitionSize: partitionSize,
+	}
+}
+
+// GetRange reads the range in parallel
+func (b *parallelBucketReader) GetRange(ctx context.Context, name string, off int64, length int64) (io.ReadCloser, error) {
+	partFilePrefix := uuid.New().String()
+	g := errgroup.Group{}
+
+	numParts := length / b.partitionSize
+	if length%b.partitionSize > 0 {
+		// A partial partition is remaining
+		numParts += 1
+	}
+
+	parts := make([]Part, 0, numParts)
+
+	partId := 0
+	for o := off; o < off+length; o += b.partitionSize {
+		l := b.partitionSize
+		if o+l > off+length {
+			// Partial partition
+			l = length - (int64(partId) * b.partitionSize)
+		}
+
+		partOff := o
+		partLength := l
+		part, err := b.createPart(partFilePrefix, partId, int(partLength))
+		if err != nil {
+			return nil, err
+		}
+		parts = append(parts, part)
+
+		g.Go(func() error {
+			rc, err := b.BucketReader.GetRange(ctx, name, partOff, partLength)
+			defer runutil.CloseWithErrCapture(&err, rc, "close object")
+			if err != nil {
+				return errors.Wrap(err, fmt.Sprintf("get range part %v", partId))
+			}
+			if _, err := io.Copy(part, rc); err != nil {
+				return errors.Wrap(err, fmt.Sprintf("get range part %v", partId))
+			}
+			part.Flush()
+			return part.Sync()
+		})
+		partId += 1
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+	return newPartMerger(parts)
+}
+
+func (b *parallelBucketReader) createPart(partFilePrefix string, partId int, size int) (Part, error) {
+	if b.tmpDir == "" {
+		// Parts stored in memory
+		return newPartBuffer(size), nil
+	}
+
+	partName := fmt.Sprintf("%s.part-%d", partFilePrefix, partId)
+	filename := filepath.Join(b.tmpDir, partName)
+	return newPartFile(filename)
+}
+
+type partMerger struct {
+	closers     []io.Closer
+	multiReader io.Reader
+}
+
+func newPartMerger(parts []Part) (*partMerger, error) {
+	readers := make([]io.Reader, 0, len(parts))
+	closers := make([]io.Closer, 0, len(parts))
+	for _, p := range parts {
+		// Seek is necessary because the part was just written to.
+		if _, err := p.Seek(0, io.SeekStart); err != nil {
```

Comment:
What if we have a `download` method on the part which will do the flush, sync and seek? This way we could reduce the surface of the interface and have less coordination.